pyracf/common/utilities/logger.py

```diff
@@ -149,30 +149,97 @@ def __redact_request_dictionary(
+        but must also contain a nested set of opened parentheses rather than a direct seqence of
+        them. The pattern looks for these nested open parenthesis as a sequence would have a )
+        character between them. Then the expression allows any non-newline characters and the
+        "end pattern" of ) and ) separated by a variable (potentially zero) whitespace.
+
+        If neither of these two patterns is found for a supplied trait_key, then it is assumed
+        this trait is set with the default pattern below.
+
+        Regex 3 ("default") - {trait_key.upper()}( +){{0,}}\(.*?(?<!\\)\)
+        This is designed to match the pattern TRAIT(value) by matching the TRAIT name, a variable
+        (potentially zero) amount of spaces, then the (value) portion which must start and end
+        with ( and ), but can conceivably contain any characters, but a negative lookbehind
+        is used to look for any escape \ character that would indicate the matching of the
+        ( and ) is otherwise a coincidence.
+
+        In all replacement expressions, the variable amounts of whitespace are captured so that
+        they can be preserved by this redaction operations.
+
         """
         asterisks = "********"
+        is_bytes = False
         if isinstance(input_string, bytes):
-            asterisks = asterisks.encode("cp1047")
-        pre_keyword = input_string[:start_ind]
-        post_keyword = end_pattern.join(input_string[start_ind:].split(end_pattern)[1:])
-        return pre_keyword + asterisks + end_pattern + post_keyword
+            input_string = input_string.decode("cp1047")
+            is_bytes = True
+        quoted = re.search(
+            rf"{trait_key.upper()}( +){{0,}}\(\'.*?(?<!\')\'\)", input_string
+        )
+        nested = re.search(
+            rf"{trait_key.upper()}( +){{0,}}\([^)]*\(.*\)( +){{0,}}\)", input_string
+        )
+        if quoted is not None:
+            input_string = re.sub(
+                rf"{trait_key.upper()}( +){{0,}}\(\'.*?(?<!\')\'\)",
+                rf"{trait_key.upper()}\1('{asterisks}')",
+                input_string,
+            )
+        else:
+            if nested is not None:
+                input_string = re.sub(
+                    rf"{trait_key.upper()}( +){{0,}}\([^)]*\(.*\)( +){{0,}}\)",
+                    rf"{trait_key.upper()}\1({asterisks})",
+                    input_string,
+                )
+            else:
+                input_string = re.sub(
+                    rf"{trait_key.upper()}( +){{0,}}\(.*?(?<!\\)\)",
+                    rf"{trait_key.upper()}\1({asterisks})",
+                    input_string,
+                )
+        if is_bytes:
+            return input_string.encode("cp1047")
+        return input_string
 
     def redact_request_xml(
         self,
         xml_string: Union[str, bytes],
         secret_traits: dict,
     ) -> Union[str, bytes]:
-        """
+        r"""
```

Comment:
Is the `r` necessary to redact the regex stuff here?

Reply:
without the "r", some of the characters used in this string like `\` take on new meaning and break the string. Treating it as a raw string prevents that sort of interpretation of special characters.